src/io/binary_parser.cpp

```diff
@@ -276,3 +289,36 @@ long long read_edges_inc(string filename, dd::FactorGraph &fg) {
   file.close();
   return count;
 }
+
+// read domains for multinomial
+void read_domains(std::string filename, dd::FactorGraph &fg) {
+  ifstream file;
+  file.open(filename.c_str(), ios::in | ios::binary);
+  long id, value;
+  long cardinality;
+  while (true) {
+    file.read((char *)&id, 8);
+    if (!file.good()) break;
+
+    id = be64toh(id);
+    dd::Variable &variable = fg.variables[id];
+    cardinality = variable.upper_bound - variable.lower_bound + 1;
+    long domain_size;
+    file.read((char *)&domain_size, 8);
+    domain_size = be64toh(domain_size);
+    assert(cardinality == domain_size);
+    variable.domain.clear();
+    variable.domain_map.clear();
+
+    for (int i = 0; i < cardinality; i++) {
+      file.read((char *)&value, 8);
+      value = be64toh(value);
+      variable.domain.push_back(value);
+    }
+
+    std::sort(variable.domain.begin(), variable.domain.end());
```

Comment:
Not sure if this sorting is necessary or actually makes grounding any easier?  Especially if we're not sorting the weight ids, things might become inconsistent once we break some implicit contract from one side.  If there's a good reason for sorting, let's clearly write down a comment here what requirements this imposes or relaxes on the input (order of the category ids, weight ids), what ordering assumptions other parts make that requires the sorting, etc.  Honestly, I'm not sure which is better: relying on positions (or giving meaning to positions) vs. the id values.